Add contrib module for HadoopFileSource.

tomwhite opened the pull request.

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,436 @@
+ * {@code
+ * HadoopFileSource<K, V> source = HadoopFileSource.from(path, MyInputFormat.class,
+ *   MyKey.class, MyValue.class);
+ * PCollection<KV<MyKey, MyValue>> records = Read.from(mySource);
```
For convenience, you may want to also provide a shorthand - HadoopFileSource.readFrom()

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,436 @@
+ * @param <V> The type of values to be read from the source.
+ */
+public class HadoopFileSource<K, V> extends BoundedSource<KV<K, V>> {
+
```
Redundant blank line?

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,436 @@
+
+  private List<InputSplit> computeSplits(long desiredBundleSizeBytes) throws IOException, IllegalAccessException,
+      InstantiationException {
+    Job job = Job.getInstance();
```
I'm not familiar with the Hadoop API - is this accessing some sort of singleton? Will it work if multiple calls to computeSplits (on different objects) are happening concurrently?

tomwhite replied:
It actually creates a new instance - I'll add a comment.

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,436 @@
+
+    @Override
+    public boolean start() throws IOException {
+      Job job = Job.getInstance();
```
Same question about this singleton-looking call.

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,436 @@
+            if (currentReader.nextKeyValue()) {
+              currentPair = nextPair();
+              return true;
+            }
```
Do you need to close currentReader 1) before line 303 2) after line 308 ?

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,436 @@
+    private KV<K, V> nextPair() throws IOException, InterruptedException {
+      K key = currentReader.getCurrentKey();
+      V value = currentReader.getCurrentValue();
+      if (key instanceof Writable) {
```
Add a comment why these need to be cloned, but only if they're Writable?

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,436 @@
+    @Override
+    public void close() throws IOException {
+      if (currentReader != null) {
+        currentReader.close();
```
Is close() on these objects idempotent? If not, set it to null.

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,436 @@
+    private Double getProgress() {
+      try {
+        return (double) currentReader.getProgress();
+      } catch (IOException e) {
```
Join the catches with | ?

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,436 @@
+
+    @Override
+    public BoundedSource<KV<K, V>> splitAtFraction(double fraction) {
+      // not supported
```
Explain in a bit more detail why it is not supported? (if it's a fundamental limitation of Hadoop file API, explain which one - e.g. so that somebody who later wants to try adding this support is aware of which limitation they're up against)

jkff commented on contrib/hadoop/src/test/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSourceTest.java:
```diff
@@ -0,0 +1,123 @@
+    }
+    assertTrue(nonEmptySplits > 2);
+  }
+
```
Some more things to test:
- That readers are closed
- That everything works correctly (including closing readers) if you read a filepattern where some of the matched files are empty
See tests in https://github.com/GoogleCloudPlatform/DataflowJavaSDK/blob/master/sdk/src/test/java/com/google/cloud/dataflow/sdk/io/FileBasedSourceTest.java - the majority are inapplicable, but some may be useful to mimic.

jkff commented on contrib/hadoop/src/test/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSourceTest.java:
```diff
@@ -0,0 +1,123 @@
+    File tmpFile = tmpFolder.newFile(filename);
+    String path = tmpFile.toString();
+    Writer writer = null;
+    try {
```
Can this use a shorter try-with-resources statement?

lukecwik commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,101 @@
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.google.cloud.dataflow</groupId>
```
Your missing the <parent> reference in this pom.xml to refer to the parent pom

It would be best if you copied the examples/pom.xml and started from there and made the edits required to it to produce the required pom.xml

lukecwik commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,101 @@
+  <artifactId>google-cloud-dataflow-java-contrib-hadoop</artifactId>
+  <name>Hadoop</name>
+  <description>Library with PTransforms for reading and writing Hadoop files.</description>
+  <version>0.0.1-SNAPSHOT</version>
```
Version should be manual_build

lukecwik commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,101 @@
+      <artifactId>google-cloud-dataflow-java-sdk-all</artifactId>
+      <version>${google-cloud-dataflow-version}</version>
+      <exclusions>
+        <!-- Use Hadoop/Spark's backend logger -->
```
Google cloud Dataflow depends on using JDK14 logger to get logs on the Dataflow workers, this comment implies that this will be incompatible with the current logging setup with the Dataflow worker.

lukecwik commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,101 @@
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.12</version>
```
Please use 4.11 so that it matches the dependencies of the other poms in the project.

lukecwik commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,101 @@
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-client</artifactId>
+      <version>2.7.0</version>
+      <scope>provided</scope>
```
By marking this as provided, this means that people will need to supply the dependency within their own project or that the container environment will make sure that its there. Why did you make this provided?

tomwhite replied:
Hadoop doesn't have great RPC client compatibility between one version and another so it's common to mark the Hadoop dependency as provided and have users specify the version they need in their project.

lukecwik commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,101 @@
+      <type>test-jar</type>
+      <scope>test</scope>
+      <exclusions>
+        <!-- Use Hadoop/Spark's backend logger -->
```
Excluding this will make the library incompatible with the Dataflow worker since it relies on the JDK14 logger being the only SLF4J backend.

lukecwik commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/WritableCoder.java:
```diff
@@ -0,0 +1,96 @@
+ *
+ * @param <T> the type of elements handled by this coder
+ */
+public class WritableCoder<T extends Writable> extends AtomicCoder<T> {
```
By extending the AtomicCoder, your saying that the encoding of all Hadoop's Writable objects is deterministic.
The Hadoop Writable interface doesn't have such a requirement. To be able to use the values as keys, Dataflow requires that the byte encoded representation is deterministic, if two values are equal, then their byte representations are equal. This is similar to the WritableComparable interface that Hadoop exposes but the big difference is that WritableComparable allows you to write your comparator while Dataflow does not.
Based upon this I would recommend that you extend StandardCoder instead of AtomicCoder.

tomwhite commented:
Thanks for the reviews @jkff and @lukecwik! I've addressed all the comments, except for the one about extra tests, which will take a bit longer.

tomwhite commented:
Looks like the API for BoundedSource has changed, which is causing the build error. I'll rebase and fix this next week.

tomwhite commented:
Rebased and added more tests. This should be ready to have another look at. @jkff @lukecwik

davorbonaci commented on pom.xml:
```diff
@@ -70,6 +70,7 @@
   <modules>
     <module>sdk</module>
     <module>examples</module>
+    <module>contrib/hadoop</module>
```
Please revert this change. We'd like to have a separate build for the `contrib` module. Perhaps it can be gated by a Maven profile in this file, or a separate aggregator `pom.xml` in contrib/ directory. It's fine to defer all that to a separate pull request, or leave that work to us.

tomwhite replied:
OK. Can you confirm that the contrib modules will be made available from Maven central?

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please add AUTHORS.md and README.md files. Please follow the pattern established by the [`join-library`](https://github.com/GoogleCloudPlatform/DataflowJavaSDK/tree/master/contrib/join-library). The README should cover all documentation needed to use the artifact.

dhalperi replied:
(Done)

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ~ Copyright (C) 2015 Google Inc.
```
You own the copyright too; let's acknowledge that by changing this line into:
`Copyright (C) 2015 The Google Cloud Dataflow Hadoop Library Authors`

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
```
I think you can drop the <parent> section.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+
+  <groupId>com.google.cloud.dataflow</groupId>
+  <artifactId>google-cloud-dataflow-java-contrib-hadoop</artifactId>
+  <name>Google Cloud Dataflow Hadoop Contrib</name>
```
Hadoop Contrib --> Hadoop Connector, or Hadoop File Source, or something similar.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+  <description>Library with PTransforms for reading and writing Hadoop files.</description>
+  <url>http://cloud.google.com/dataflow</url>
+
+  <version>manual_build</version>
```
Optionally, if you like notation 0.0.1-SNAPSHOT better, feel free to use it.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+
+  <packaging>jar</packaging>
+
+  <profiles>
```
Please drop profiles section; unused at the moment.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+        <artifactId>maven-compiler-plugin</artifactId>
+      </plugin>
+
+      <plugin>
```
Seems unused, please drop.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+      </plugin>
+
+      <!-- Source plugin for generating source and test-source JARs. -->
+      <plugin>
```
Seems unused, please drop.

tomwhite replied:
This generates source jars, which are useful for IDEs (so you can click through to see the source).

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+        </executions>
+      </plugin>
+
+      <plugin>
```
Seems unused, please drop.

tomwhite replied:
Ditto - useful for IDEs.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+        </executions>
+      </plugin>
+
+      <plugin>
```
Please drop.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+      </plugin>
+
+      <!-- Coverage analysis for unit tests. -->
+      <plugin>
```
Seems unused, please drop.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+        </executions>
+      </plugin>
+
+      <plugin>
```
Probably inferred by default and not necessary. Please drop.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+    <dependency>
+      <groupId>com.google.cloud.dataflow</groupId>
+      <artifactId>google-cloud-dataflow-java-sdk-all</artifactId>
+      <version>${project.version}</version>
```
Please follow the pattern established by the `join-library`. Add a property `google-cloud-dataflow-version`. (This way the module can be revisioned separately.)

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+    <dependency>
+      <groupId>com.google.guava</groupId>
+      <artifactId>guava</artifactId>
+      <version>18.0</version>
```
FYI: We have had interoperability issues with Hadoop and Guava 18 in the past. This might be a source of issues.

tomwhite replied:
Guava was not actually being used by this module, so I removed it.

davorbonaci commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,245 @@
+    <dependency>
+      <groupId>com.google.cloud.dataflow</groupId>
+      <artifactId>google-cloud-dataflow-java-sdk-all</artifactId>
+      <version>${project.version}</version>
```
Depend on a version from a property, as above.

davorbonaci commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,484 @@
+    this.serializableSplit = serializableSplit;
+  }
+
+  public String getFilepattern() {
```
Optionally, javadoc on all public getters.

davorbonaci commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/WritableCoder.java:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @param <T> the type of elements handled by this coder
+ */
+public class WritableCoder<T extends Writable> extends StandardCoder<T> {
```
You may want to consider extending `CustomCoder`.

davorbonaci replied:
You may want to test the coder. Use `CoderProperties`.

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,484 @@
+    long size = 0;
+    try {
+      Job job = Job.getInstance(); // new instance
+      for (FileStatus st : listStatus(createFormat(job), job)) {
```
Note that if there are really a lot of files, this is going to take a long time. The code will benefit from 1) parallelism and 2) sampling - feel free to mimic something like FileBasedSource does in https://github.com/GoogleCloudPlatform/DataflowJavaSDK/commit/c6f488bb3d57f3f19610e86c283e9035d0f22f42
(ok to leave as is - it can be addressed later)

jkff replied:
Oh, looks like you're not doing any IO in the loop, all the work is done by listStatus which I suppose is already optimized.

jkff commented on contrib/hadoop/src/test/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSourceTest.java:
```diff
@@ -0,0 +1,189 @@
+  @Test
+  public void testFullyReadSingleFile() throws Exception {
+    PipelineOptions options = PipelineOptionsFactory.create();
+    List<KV<IntWritable, Text>> expectedResults = createRandomRecords(3, 10000, 0);
```
Any reason for choosing 10000 records rather than a smaller number (for faster tests) - is 10000 necessary to trigger some more interesting behavior?

tomwhite replied:
10000 is needed for `testSplits`, but not for the others, so I changed them to use 10 records.

davorbonaci commented:
Hi Tom,
Thanks for this contribution. The code looks quite good -- we'd be happy to merge it.

I have left ~20 comments, mostly around the process and the `pom.xml` file -- most of them are very trivial. @jkff is also taking a quick peek.

You may also want to squash commits into a single one.

Thanks!

jkff commented:
Thanks! I added a couple more minor comments, but overall this is looking very good and I'm happy to get this in.

tomwhite commented:
@davorbonaci Thanks for the comments! I based the POM on the one for the examples module based on an earlier comment, but I've now changed it in the way you suggested so it's in line with the other contrib module. I've also addressed the other comments.

@jkff Thanks - hopefully this is ready to go now. I've squashed into a single commit.

jkff commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,484 @@
+              currentPair = nextPair();
+              return true;
+            }
+            currentReader.close();
```
Should you set it to null so line 344 on the next iteration doesn't close it again? (if .close() is idempotent, fine to leave as is)

tomwhite replied:
Fixed. (Although `close` should be idempotent, it's fine to set to null to avoid calling it twice.)

jkff commented:
Looks great. I added a single comment that may be fine to ignore.

dhalperi commented:
@tomwhite I apologize profusely for the delay here. I'm in the process of finding and picking up the balls we've dropped.

@jkff you've done initial review and approve?

Since this is my first time seeing this code, I will take a minor review pass tomorrow.

Thanks, Dan

dhalperi commented on contrib/hadoop/pom.xml:
```diff
@@ -0,0 +1,175 @@
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <google-cloud-dataflow-version>manual_build</google-cloud-dataflow-version>
```
this is probably the wrong version? Maybe `[1.0.0,2.0.0)`?

dhalperi commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,485 @@
+ * }
+ * </pre>
+ *
+ * Implementation note: Since Hadoop's {@link org.apache.hadoop.mapreduce.lib.input.FileInputFormat}
```
New `<p>`?

dhalperi commented on contrib/hadoop/README.md:
```diff
@@ -0,0 +1,21 @@
+transform. For example:
+
+    PCollection<KV<MyKey, MyValue>> records = HadoopFileSource.readFrom(path,
+      MyInputFormat.class, MyKey.class, MyValue.class);
```
If it is true that this reader is mainly for Hadoop SequenceFiles, it might be nice to add a bit more to the documentation or make the documentation more precise.

I thought based on the name that this Source might read arbitrary record types from HDFS, etc. (But of course you're the Hadoop expert, so you would know better whether this is clear enough.)

weiminyu-personal commented on contrib/hadoop/src/main/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSource.java:
```diff
@@ -0,0 +1,485 @@
+
+  @SuppressWarnings("unchecked")
+  private <T> Coder<T> getDefaultCoder(Class<T> c) {
+    if (Writable.class.isAssignableFrom(c)) {
```
SequenceFile exported by Hbase has <Writable, Result> pairs. Result is not a Writable implementor. 
We may want to allow user to register customer coders in this class.
